eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala

```diff
@@ -73,485 +75,270 @@ object ReplaceableTxPublisher {
-                case u: UpdateFulfillHtlc if Crypto.sha256(u.paymentPreimage) == claimHtlcSuccess.paymentHash => u.paymentPreimage
-              }
-              preimage_opt match {
-                case Some(preimage) =>
-                  val signedClaimHtlcTx = addSigs(claimHtlcSuccess, sig, preimage)
-                  publish(replyTo, cmd, signedClaimHtlcTx.tx, fee)
-                case None =>
-                  log.error("preimage not found for htlcId={}, skipping...", claimHtlcSuccess.htlcId)
-                  sendResult(replyTo, TxPublisher.TxRejected(loggingInfo.id, cmd, TxPublisher.TxRejectedReason.TxSkipped(retryNextBlock = false)))
-              }
-            case claimHtlcTimeout: ClaimHtlcTimeoutTx =>
-              val signedClaimHtlcTx = addSigs(claimHtlcTimeout, sig)
-              publish(replyTo, cmd, signedClaimHtlcTx.tx, fee)
+      case WrappedTxResult(MempoolTxMonitor.TxConfirmed(confirmedTx)) => sendResult(replyTo, TxPublisher.TxConfirmed(cmd, confirmedTx))
+      case WrappedTxResult(MempoolTxMonitor.TxRejected(_, reason)) =>
+        replyTo ! TxPublisher.TxRejected(loggingInfo.id, cmd, reason)
+        // We wait for our parent to stop us: when that happens we will unlock utxos.
+        Behaviors.same
+      case WrappedCurrentBlockCount(currentBlockCount) =>
+        // We avoid a herd effect whenever a new block is found.
+        timers.startSingleTimer(CheckFeeKey, CheckFee(currentBlockCount), (1 + Random.nextLong(nodeParams.maxTxPublishRetryDelay.toMillis)).millis)
+        Behaviors.same
+      case CheckFee(currentBlockCount) =>
+        val currentFeerate = getFeerate(nodeParams.onChainFeeConf.feeEstimator, cmd.txInfo.confirmBefore, BlockHeight(currentBlockCount))
+        val targetFeerate_opt = if (cmd.txInfo.confirmBefore.toLong <= currentBlockCount + 6) {
+          log.debug("{} confirmation target is close (in {} blocks): bumping fees", cmd.desc, cmd.txInfo.confirmBefore.toLong - currentBlockCount)
+          // We make sure we increase the fees by at least 20% as we get close to the confirmation target.
+          Some(currentFeerate.max(tx.feerate * 1.2))
+        } else if (tx.feerate * 1.2 <= currentFeerate) {
+          log.debug("{} confirmation target is in {} blocks: bumping fees", cmd.desc, cmd.txInfo.confirmBefore.toLong - currentBlockCount)
+          Some(currentFeerate)
+        } else {
+          log.debug("{} confirmation target is in {} blocks: no need to bump fees", cmd.desc, cmd.txInfo.confirmBefore.toLong - currentBlockCount)
+          None
+        }
+        targetFeerate_opt.foreach(targetFeerate => {
+          // We check whether our currently published transaction is confirmed: if it is, it doesn't make sense to bump
+          // the fee, we're just waiting for enough confirmations to report the result back.
+          context.pipeToSelf(bitcoinClient.getTxConfirmations(tx.signedTx.txid)) {
```

Comment:
> (btw, there is no herd effect protection in the MempoolTxMonitor).

I don't think we need it, do we? You shouldn't have thousands of `MempoolTxMonitor` running at the same time, waiting for confirmations, since this is limited by the number of channels in a force-close state you have.

> have the MempoolTxMonitor notify the ReplaceableTxPublisher after each block and subsequent call to getTxConfirmations with a NotYetConfirmed message that the publisher uses to evaluate whether we should bump instead of the (WrappedCurrentBlockCount/CheckFee combo)

I'll try that. My first instinct is that it forces more knowledge on the `ReplaceableTxPublisher` than necessary, because it needs to know more "internal" steps of the `MempoolTxMonitor` and keep state accordingly (whereas here just checking one pre-condition feels simple enough).

Reply:
> I don't think we need it, do we? You shouldn't have thousands of MempoolTxMonitor running at the same time, waiting for confirmations, since this is limited by the number of channels in a force-close state you have.

Then we don't need the herd effect protection in `ReplaceableTxPublisher` either right? I think it makes sense to have that kind of mechanism, because we want this to work well even under a catastrophic mass force-close scenario.

Reply:
Actually the one in `ReplaceableTxPublisher` isn't for the herd effect, it's my bad that I didn't update the comment.
It's to give time to the underlying `MempoolTxMonitor` to handle the new block first, and potentially report that the tx is confirmed (or evicted) before we make a decision to fee-bump (or not) inside the `ReplaceableTxPublisher`.

But your comments make me realize that it would be cleaner to rely on a signal from the `MempoolTxMonitor` to decide that. Instead of having the `ReplaceableTxPublisher` register to new block events, what you suggested about having the `MempoolTxMonitor` tell us the state of the tx at each block makes sense: then we can decide to bump or not. I'll dive into this.

> I think it makes sense to have that kind of mechanism, because we want this to work well even under a catastrophic mass force-close scenario.

That sounds fair, I'll add this to the `MempoolTxMonitor` then!

Reply:
Done in https://github.com/ACINQ/eclair/pull/2113/commits/8d13b911e5c6743bdc9c51909e9901bc093adb04, that was a good idea, it's cleaner now.